### PR TITLE
[Feat] support layerwise offload for Bagel

### PR DESCRIPTION
**Pull request by lsyyysky**
## Purpose
Add Bagel layerwise CPU offload support.
This PR resolves the Bagel gap in CPU Offload (Layerwise) support from RFC #1217 by:
- adding Bagel-specific offload module discovery so the backend can distinguish block modules, VAE, and resident modules
- moving Bagel resident modules to GPU while applying layerwise hooks only to the LLM transformer blocks
- avoiding the eager `self.to(self.device)` path in `BagelPipeline` when layerwise offload is enabled
## Test Result
vllm serve --model /satadata/Bagel/BAGEL-7B-MoT/ --omni --enable_layerwise_offload --port 8004

python3 examples/online_serving/bagel/openai_chat_client.py   --server http://localhost:8004   --modality text2img   --prompt "a majestic dragon perched on the mountain ridge of Vermont, misty morning atmosphere, photorealistic style"   --height 1024   --width 1024   --steps 50   --seed 42   --output docs/examples/dragon_chat.png

Offload VS no offload

| Model | No Offload | No Offload GPU Mem | With Offload | With Offload GPU Mem |
| -- | -- | -- | -- | -- |
| BAGEL-7B-MoT | image | 63.8 GB | image| 44.8 GB |

### CODE REVIEW

**Comment by princepride** [comment]
@lsyyysky You may need refer to this doc: https://docs.vllm.ai/en/stable/contributing/#pull-requests-code-reviews to sign off and ensure coding styles.

**Comment by lsyyysky** [comment]
> @lsyyysky You may need refer to this doc: https://docs.vllm.ai/en/stable/contributing/#pull-requests-code-reviews to sign off and ensure coding styles.

LGTM

**Comment by hsliuustc0106** [comment]
PR #2734 - [Feat] support layerwise offload for Bagel

OVERALL: NO BLOCKERS
VERDICT: COMMENT

Correctness: PASS
Reliability: PASS
Breaking: PASS
Tests: PASS (GPU mem: 63.8 GB -> 44.8 GB)
Documentation: PASS (inline doc added)
Security: PASS

Summary: Bagel layerwise CPU offload with --cpu-offload-bagel option. 48 add, 2 del. Measurements: H100x8, Wan2.2-14B-T2V-720P, GPU mem 63.8 GB -> 44.8 GB. Working feature, no blockers.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 1]
```diff
@@ -915,3 +814,31 @@ def _filtered_weights():
 
         loader = AutoWeightsLoader(self)
         return loader.load_weights(_filtered_weights())
+
+    def discover_offload_modules(self) -> PipelineModules:
```
Comment (wtomin) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: Better to update `docs/user_guide/diffusion/cpu_offload_diffusion.md` and use Bagel as an example for how to support layerwise cpu offloading for customized modules.

**Comment by wtomin** [comment]
1. Please update `docs/user_guide/diffusion_features.md` to reflect Bagel's CPU offload support in the feature matrix table.
2.  Suggest to update `tests/e2e/online_serving/test_bagel_expansion.py` to include the layerwise offloading feature test in nightly CI. I will add a `nightly-test` label to launch buildkite test when it's ready.

**Comment by lsyyysky** [comment]
> 1. Please update `docs/user_guide/diffusion_features.md` to reflect Bagel's CPU offload support in the feature matrix table.
> 2. Suggest to update `tests/e2e/online_serving/test_bagel_expansion.py` to include the layerwise offloading feature test in nightly CI. I will add a `nightly-test` label to launch buildkite test when it's ready.

Thanks for your suggest! Finished.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py (hunk 1 above)
Comment (lsyyysky) [reply vllm_omni/diffusion/models/bagel/pipeline_bagel.py]: Finished

**Comment by princepride** [comment]
@yuanheng-zhao PTAL

**Comment by princepride** [comment]
@wtomin why nightly-test failed?

File: vllm_omni/diffusion/offloader/module_collector.py [hunk 2]
```diff
@@ -35,6 +37,14 @@ def discover(pipeline: nn.Module) -> PipelineModules:
         Returns:
             PipelineModules with lists of discovered modules and names
         """
+        # Custom discovery
+        custom_discover = getattr(pipeline, "discover_offload_modules", None)
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/module_collector.py]: This is conflicting with the functionality provided in https://github.com/vllm-project/vllm-omni/pull/2427 
We'd better keep only one of them @lsyyysky  @NickCao

Comment (lsyyysky) [reply vllm_omni/diffusion/offloader/module_collector.py]: Yeah, but i think the custom discovery need to be implemented in #2427. It does not yet support the nested path needed by Bagel here (for example, language_model.model), nor does it cover resident_modules @yuanheng-zhao @NickCao

**Review by lishunyang12 — COMMENTED** [review]
## Review: [Feat] support layerwise offload for Bagel

Overall this is a clean, well-structured PR that follows the existing offload patterns nicely. The custom `discover_offload_modules()` extension point is a good design choice for non-standard pipelines. A few items to address:

### Issues

1. **Documentation code snippet has an undefined variable** (`docs/user_guide/diffusion/cpu_offload_diffusion.md`)

   The example `discover_offload_modules()` in the doc references `resident_names` in the return statement, but this variable is never defined in the snippet. The actual implementation in `pipeline_bagel.py` correctly builds `resident_names` via the `add_resident` helper, but the doc example is misleading and would not work if copy-pasted. Consider either adding the construction of `resident_names` to the snippet or simplifying the example to show the pattern clearly.

2. **Redundant parentheses in condition** (`pipeline_bagel.py`, line ~277)

   ```python
   if quant_config is None and not (od_config.enable_layerwise_offload):
   ```

   The parentheses around `od_config.enable_layerwise_offload` are unnecessary. Minor, but:
   ```python
   if quant_config is None and not od_config.enable_layerwise_offload:
   ```

3. **Missing error handling for resident module GPU moves** (`layerwise_backend.py`)

   The VAE move to GPU is wrapped in a try/except, but the new resident module moves are not:
   ```python
   for name, module in zip(modules.resident_names, modules.resident_modules):
       module.to(self.device)
   ```
   If any resident module fails to move (e.g., meta tensors from quantization, or unexpected state), this will crash the entire offload setup. Consider wrapping in try/except with a warning, consistent with the VAE handling above it.

4. **Resident modules moved *after* the log line but *before* hook application** -- ordering is correct, just noting it is intentional and good. The resident modules land on GPU before the block hooks are applied.

### Suggestions (non-blocking)

5. **`discover_offload_modules` could validate that `self.bagel` and `self.language_model` exist** before accessing sub-attributes. Currently, if the pipeline is partially initialized or `bagel` is None, this will raise an opaque `AttributeError`. A guard or a clearer error message would help debugging.

6. **Consider adding `self.bagel.language_model.lm_head`** (or `embed_tokens`) to the resident modules list if it exists. These are typically small and accessed every forward pass. Currently they would be treated as non-block children of the DiT and moved to GPU by the generic loop at line 337-341 of `layerwise_backend.py`, so this is fine in practice -- just calling it out for awareness.

7. **f-string in logging** (`layerwise_backend.py` line ~309): The new log uses f-string (`logger.debug(f"Moving resident module {name}...")`), while surrounding code uses `%s` style. Minor style inconsistency.

### Positive notes

- The `PipelineModules` dataclass extension with `resident_modules`/`resident_names` using `field(default_factory=list)` is backward-compatible and clean.
- The custom discovery hook in `ModuleDiscovery.discover()` with the `callable` check and `TypeError` guard is well-done.
- Test coverage via the e2e expansion test is good.
- The GPU memory savings (63.8 GB -> 44.8 GB) with visual parity is a strong result.

Items 1 and 3 are worth fixing before merge. The rest are minor or optional.

File: docs/user_guide/diffusion/cpu_offload_diffusion.md [hunk 3]
```diff
@@ -104,6 +104,53 @@ class Flux2Transformer2DModel(nn.Module):
     _layerwise_offload_blocks_attrs = ["transformer_blocks", "single_transformer_blocks"]
 ```
 
+#### Customized Pipelines
+For customized pipelines that do not follow the standard `transformer` / `text_encoder` / `vae`
+layout, the pipeline can provide a custom `discover_offload_modules()` implementation.
+
+Bagel is one such example. Its transformer blocks live under `language_model.model`, while
+several Bagel-specific modules (such as `time_embedder`, `vae2llm`, `llm2vae`, `vit_model`, etc.)
+should remain resident on GPU during layerwise offloading. Register
```
Comment (yuanheng-zhao) [inline docs/user_guide/diffusion/cpu_offload_diffusion.md]: Modules like `time_embedder` and `vit_model` seem not to be included in layer modules? I think layerwise offloading won't move those components out of scope of layer module.

File: vllm_omni/diffusion/offloader/module_collector.py (hunk 2 above)
Comment (yuanheng-zhao) [reply vllm_omni/diffusion/offloader/module_collector.py]: Is `resident_modules` aiming at keeping specific modules on device? For module outside layer modules specified in `_layerwise_offload_blocks_attrs`, layerwise offloading won't move it back and forth

File: docs/user_guide/diffusion/cpu_offload_diffusion.md (hunk 3 above)
Comment (lsyyysky) [reply docs/user_guide/diffusion/cpu_offload_diffusion.md]: 1、Yes, you are right. These modules are indeed not part of the layer modules. I can delete "during layerwise offloading".
2、Yse, resident_modules is meant to explicitly keep such specific  modules on device.

**Comment by lsyyysky** [comment]
> ## Review: [Feat] support layerwise offload for Bagel
> Overall this is a clean, well-structured PR that follows the existing offload patterns nicely. The custom `discover_offload_modules()` extension point is a good design choice for non-standard pipelines. A few items to address:
> 
> ### Issues
> 1. **Documentation code snippet has an undefined variable** (`docs/user_guide/diffusion/cpu_offload_diffusion.md`)
>    The example `discover_offload_modules()` in the doc references `resident_names` in the return statement, but this variable is never defined in the snippet. The actual implementation in `pipeline_bagel.py` correctly builds `resident_names` via the `add_resident` helper, but the doc example is misleading and would not work if copy-pasted. Consider either adding the construction of `resident_names` to the snippet or simplifying the example to show the pattern clearly.
> 2. **Redundant parentheses in condition** (`pipeline_bagel.py`, line ~277)
>    ```python
>    if quant_config is None and not (od_config.enable_layerwise_offload):
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    The parentheses around `od_config.enable_layerwise_offload` are unnecessary. Minor, but:
>    ```python
>    if quant_config is None and not od_config.enable_layerwise_offload:
>    ```
> 3. **Missing error handling for resident module GPU moves** (`layerwise_backend.py`)
>    The VAE move to GPU is wrapped in a try/except, but the new resident module moves are not:
>    ```python
>    for name, module in zip(modules.resident_names, modules.resident_modules):
>        module.to(self.device)
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    If any resident module fails to move (e.g., meta tensors from quantization, or unexpected state), this will crash the entire offload setup. Consider wrapping in try/except with a warning, consistent with the VAE handling above it.
> 4. **Resident modules moved _after_ the log line but _before_ hook application** -- ordering is correct, just noting it is intentional and good. The resident modules land on GPU before the block hooks are applied.
> 
> ### Suggestions (non-blocking)
> 5. **`discover_offload_modules` could validate that `self.bagel` and `self.language_model` exist** before accessing sub-attributes. Currently, if the pipeline is partially initialized or `bagel` is None, this will raise an opaque `AttributeError`. A guard or a clearer error message would help debugging.
> 6. **Consider adding `self.bagel.language_model.lm_head`** (or `embed_tokens`) to the resident modules list if it exists. These are typically small and accessed every forward pass. Currently they would be treated as non-block children of the DiT and moved to GPU by the generic loop at line 337-341 of `layerwise_backend.py`, so this is fine in practice -- just calling it out for awareness.
> 7. **f-string in logging** (`layerwise_backend.py` line ~309): The new log uses f-string (`logger.debug(f"Moving resident module {name}...")`), while surrounding code uses `%s` style. Minor style inconsistency.
> 
> ### Positive notes
> * The `PipelineModules` dataclass extension with `resident_modules`/`resident_names` using `field(default_factory=list)` is backward-compatible and clean.
> * The custom discovery hook in `ModuleDiscovery.discover()` with the `callable` check and `TypeError` guard is well-done.
> * Test coverage via the e2e expansion test is good.
> * The GPU memory savings (63.8 GB -> 44.8 GB) with visual parity is a strong result.
> 
> Items 1 and 3 are worth fixing before merge. The rest are minor or optional.

Thanks for your good suggestions. 1,2,3,5,7 Finished

**Comment by lsyyysky** [comment]
@princepride @wtomin @yuanheng-zhao  @lishunyang12 Could it be merged?

**Comment by yuanheng-zhao** [comment]
To find out nested modules, it will be more explicit and convenient to have some rules assigned as `A.B.target_module`, such like
```python
_dit_modules: ClassVar[list[str]] = ["pipe.transformer"]
_resident_modules: ClassVar[list[str]] = ["bagel.time_embedder", ...]
```

I just saw @NickCao had updated #2427 to include this functionality. Would you mind rebasing on @NickCao 's branch so that we keep a single protocol rather than both `SupportsModuleOffload` and `discover_offload_modules`

**Comment by lsyyysky** [comment]
> To find out nested modules, it will be more explicit and convenient to have some rules assigned as `A.B.target_module`, such like
> 
> ```python
> _dit_modules: ClassVar[list[str]] = ["pipe.transformer"]
> _resident_modules: ClassVar[list[str]] = ["bagel.time_embedder", ...]
> ```
> 
> I just saw @NickCao had updated #2427 to include this functionality. Would you mind rebasing on @NickCao 's branch so that we keep a single protocol rather than both `SupportsModuleOffload` and `discover_offload_modules`

i don't mind.

**Comment by lsyyysky** [comment]
Rebased @yuanheng-zhao

**Comment by yuanheng-zhao** [comment]
PTAL @NickCao 
The order of merging #2427 -> #2734

**Comment by lsyyysky** [comment]
> PTAL @NickCao The order of merging #2427 -> #2734
Could it be merged?

**Comment by lishunyang12** [comment]
Resolve CI failure.

**Comment by lsyyysky** [comment]
> Resolve CI failure.

I feel the CI failure is unrelated to my PR. @lishunyang12 @yuanheng-zhao

**Review by lishunyang12 — APPROVED** [review]

**Comment by lsyyysky** [comment]
Same CI failure @lishunyang12 @hsliuustc0106
